eclipse/src/saros/filesystem/EclipseReferencePointManager.java

```diff
@@ -0,0 +1,215 @@
+package saros.filesystem;
+
+import java.util.concurrent.ConcurrentHashMap;
+import org.eclipse.core.resources.IFile;
+import org.eclipse.core.resources.IFolder;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.IResource;
+import org.eclipse.core.runtime.IPath;
+import saros.activities.SPath;
+
+/**
+ * The EclipseReferencePointManager maps an {@link IReferencePoint} reference point to {@link
+ * IProject} project
+ */
+public class EclipseReferencePointManager {
+
+  ConcurrentHashMap<IReferencePoint, org.eclipse.core.resources.IProject>
+      referencePointToProjectMapper;
+
+  public EclipseReferencePointManager() {
+    referencePointToProjectMapper =
+        new ConcurrentHashMap<IReferencePoint, org.eclipse.core.resources.IProject>();
+  }
+
+  /**
+   * Creates and returns the {@link IReferencePoint} reference point of given {@link IResource}
+   * resource. The reference point points on the resource's project full path.
+   *
+   * @param resource for which the reference point should be created
+   * @return the reference point of given resourcen
+   */
+  public static IReferencePoint create(IResource resource) {
+    if (resource == null) return null;
+
+    IProject project = resource.getProject();
+    saros.filesystem.IPath path = ResourceAdapterFactory.create(project.getFullPath());
+
+    return new ReferencePointImpl(path);
+  }
+
+  /**
+   * Inserts a pair of {@link IReferencePoint} reference point and {@link IProject} project. The
+   * reference point will be created from the project's relative path by
+   * {#putIfAbsent(IReferencePoint referencePoint, IProject project) putIfAbsent} .
+   *
+   * @param project the {@link IProject} project for which a reference point is created and are
+   *     inserted.
+   */
+  public void putIfAbsent(IProject project) {
+    IReferencePoint referencePoint = create(project);
+
+    putIfAbsent(referencePoint, project);
+  }
+
+  /**
+   * Insert a pair of {@link IReferencePoint} reference point and {@link IProject} project
+   *
+   * @param referencePoint the key of the pair
+   * @param project the value of the pair
+   */
+  public void putIfAbsent(IReferencePoint referencePoint, IProject project) {
```

Comment:
The class `ConcurrentHashMap` already supports the method `putIfAbsent` which should be used. Your solution is not save, because the "contains-check + put" is not an atomic operation.

Reply:
fixed